articles/121_realtime_proposal.md

```diff
@@ -0,0 +1,227 @@
+Safety-critical systems often require a hard real-time with system with high predictability.
+
+### Implementation choices: computer architecture and OS
+
+Processors vary in capability, number of cores, memory size, and instruction set architectures, some which are better suited to real-time performance than others.
+The selection of computer architecture may be due to cost, size, or energy efficiency, which could set a constraint for the rest of the system.
+A single robot may contain multiple computers with a variety of processor types, from the main control board to sensors.
+
+Another major consideration when designing a real-time system is the choice of operating system.
+A true "hard" real-time operating system is truly deterministic.
+"Soft" real-time operating systems are designed to meet most deadlines with low jitter, but are nondeterministic.
+Most conventional user-facing operating systems are not designed with fairness and predictability as the first priorities, often because they have to support graphical displays and interactivity.
+Here are some features critical to predictable performance that are often exposed to the user in an RTOS, but not in a conventional OS (see Wikipedia, [Real-time operating system](https://en.wikipedia.org/wiki/Real-time_operating_system)).
+For more information about why these features matter, refer to the [introductory article on real-time systems](http://design.ros2.org/articles/realtime_background.html).
+
+- Task scheduling
+
+  - Expose thread priority and deterministic schedulers that incorporate priority
+  - Allow full thread preemption
+
+- Concurrency and synchronization primitives
+
+  - Floored semaphores (allow a higher priority thread to take control of locked semaphore)
+
+- Interrupt handling
+
+  - Disable interrupts
+  - Mask interrupts (temporarily suspend interrupts in a critical section)
+
+- Memory allocation
+
+Some real-time systems have a non real-time component, e.g. for graphical display or other user interaction.
+This would require a soft real-time operating system that allows non real-time threads to be scheduled alongside real-time threads.
+
+Another important constraint is which processor architectures are supported by which operating systems, and which RTOS is an appropriate choice due to performance constraints of the architecture.
+
+## Goal for ROS 2
+
+In order to support real-time systems in ROS 2, we propose an iterative approach for identifying limitations in the API that force the code path into nondeterministic/unpredictable behavior and performance bottlenecks.
+
+1. Select two use cases with real-time requirements that represent two extremes of the space described above, for example:
+
+   - An STM32 embedded board running no operating system or a hard real-time OS with 1 kHZ update frequency with 3% allowable jitter.
```

Comment:
It's an arbitrary choice. I could write a bit more about how these requirements are chosen, but in this case the requirement doesn't matter as much as profiling the jitter and driving it down does. Any advice on this matter?